databricks/koalas/frame.py

```diff
@@ -2478,16 +2488,206 @@ def transform(self, func):
+        """
+        from databricks.koalas.groupby import GroupBy
+        from databricks.koalas import Series
+
+        assert callable(func), "the first argument should be a callable function."
+        spec = inspect.getfullargspec(func)
+        return_sig = spec.annotations.get("return", None)
+        should_infer_schema = return_sig is None
+
+        if should_infer_schema:
+            # Here we execute with the first 1000 to get the return type.
+            # If the records were less than 1000, it uses pandas API directly for a shortcut.
+            limit = get_option("compute.shortcut_limit")
+            pdf = self.head(limit + 1)._to_internal_pandas()
+            transformed = func(pdf)
+            if not isinstance(transformed, (pd.DataFrame, pd.Series)):
+                raise ValueError(
+                    "The given function should return a frame; however, "
+                    "the return type was %s." % type(transformed)
+                )
+            if len(transformed) != len(pdf):
+                raise ValueError("transform_batch cannot produce aggregated results")
+            kdf_or_kser = ks.from_pandas(transformed)
+
+            if isinstance(kdf_or_kser, ks.Series):
+                kser = kdf_or_kser
+                pudf = pandas_udf(
+                    func, returnType=kser.spark_type, functionType=PandasUDFType.SCALAR
+                )
+                columns = self._internal.spark_columns
+                # TODO: Index will be lost in this case.
+                internal = self._internal.copy(
+                    spark_column=pudf(F.struct(*columns)),
+                    column_labels=kser._internal.column_labels,
+                    column_label_names=kser._internal.column_label_names,
+                )
+                return Series(internal, anchor=self)
+            else:
+                kdf = kdf_or_kser
+                if len(pdf) <= limit:
+                    # only do the short cut when it returns a frame to avoid
+                    # operations on different dataframes in case of series.
+                    return kdf
+                return_schema = kdf._internal.to_internal_spark_frame.schema
+                # Force nullability.
+                return_schema = StructType(
+                    [StructField(field.name, field.dataType) for field in return_schema.fields]
+                )
+                output_func = GroupBy._make_pandas_df_builder_func(
+                    self, func, return_schema, retain_index=True
+                )
+                pudf = pandas_udf(
+                    output_func, returnType=return_schema, functionType=PandasUDFType.SCALAR
+                )
+                temp_struct_column = verify_temp_column_name(
+                    self._internal.spark_frame, "__temp_struct__"
+                )
+                columns = self._internal.spark_columns
+                applied = pudf(F.struct(*columns)).alias(temp_struct_column)
```

Comment:
`pandas_udf` in `Spark<3` doesn't support struct type as input/output.

Reply:
damn ...

Thanks for comments, I will try to address them.